Training and Finetuning Embedding Models with Sentence Transformers v3 (#2104)
huggingface/blog: train-sentence-transformers.md

- "Apple launches the new iPad"
- "NVIDIA is gearing up for the next GPU generation"

Depending on the use case, we might want similar or dissimilar embeddings for these texts. For instance, a classification model for news articles could treat these texts as similar since they both belong to the Technology category. On the other hand, a semantic textual similarity or retrieval model should consider them dissimilar due to their distinct meanings.

## Training Components

Training Sentence Transformer models involves the following components:

1. [**Dataset**](#dataset): The data used for training and evaluation.
2. [**Loss Function**](#loss-function): A function that quantifies the model's performance and guides the optimization process.
3. [**Training Arguments**](#training-arguments) (optional): Parameters that influence training performance and tracking/debugging.
4. [**Evaluator**](#evaluator) (optional): A tool for evaluating the model before, during, or after training.
5. [**Trainer**](#trainer): Brings together the model, dataset, loss function, and other components for training.

Now, let's dive into each of these components in more detail.

## Dataset

> The [`SentenceTransformerTrainer`](https://sbert.net/docs/package_reference/sentence_transformer/SentenceTransformer.html#sentence_transformers.SentenceTransformer) uses [`datasets.Dataset`](https://huggingface.co/docs/datasets/main/en/package_reference/main_classes#datasets.Dataset) or [`datasets.DatasetDict`](https://huggingface.co/docs/datasets/main/en/package_reference/main_classes#datasets.DatasetDict) instances for training and evaluation. You can load data from the Hugging Face Datasets Hub or use local data in various formats such as CSV, JSON, Parquet, Arrow, or SQL.

Reviewer: The first link is still not available, I suppose it's pending release.

Author: Yes, almost all links are currently inaccessible as they rely on the new documentation structure. For reference, I'm splitting the documentation into the model type:
* Sentence Transformer (a.k.a bi-encoder)
* Cross Encoder (a.k.a reranker)

The latter isn't very well supported yet, but I intend to expand on it in the future. This restructuring allows for more model types in the future, like sparse models and ColBERT models.
I've already written the redirection rules to prevent 404s when users are pointed to the "old" documentation pages that won't exist at those locations anymore.

Note: Many Hugging Face datasets that work out of the box with Sentence Transformers have been tagged with `sentence-transformers`, allowing you to easily find them by browsing to [https://huggingface.co/datasets?other=sentence-transformers](https://huggingface.co/datasets?other=sentence-transformers). We strongly recommend that you browse these datasets to find training datasets that might be useful for your tasks.

### Data on Hugging Face Hub

> To load data from the Hugging Face Datasets Hub, use the [`datasets.load_dataset`](https://huggingface.co/docs/datasets/main/en/package_reference/loading_methods#datasets.load_dataset) function:

Reviewer: ```suggestion
To load data from datasets in the Hugging Face Hub, use the [`load_dataset`](https://huggingface.co/docs/datasets/main/en/package_reference/loading_methods#datasets.load_dataset) function:
```

```python
from datasets import load_dataset

train_dataset = load_dataset("sentence-transformers/all-nli", "pair-class", split="train")
eval_dataset = load_dataset("sentence-transformers/all-nli", "pair-class", split="dev")

print(train_dataset)
"""
Dataset({
    features: ['premise', 'hypothesis', 'label'],
    num_rows: 942069
})
"""
```

Some datasets, like [`sentence-transformers/all-nli`](https://huggingface.co/datasets/sentence-transformers/all-nli), have multiple subsets with different data formats. You need to specify the subset name along with the dataset name.

### Local Data (CSV, JSON, Parquet, Arrow, SQL)

> If you have local data in common file formats, you can easily load it using [`datasets.load_dataset`](https://huggingface.co/docs/datasets/main/en/package_reference/loading_methods#datasets.load_dataset):

Reviewer: ```suggestion
If you have local data in common file formats, you can easily load it using [`load_dataset` too](https://huggingface.co/docs/datasets/main/en/package_reference/loading_methods#datasets.load_dataset):
```

Author: I moved the "too" out of the URL, otherwise this look good, will accept it in a minute.

```python
from datasets import load_dataset

dataset = load_dataset("csv", data_files="my_file.csv")
# or
dataset = load_dataset("json", data_files="my_file.json")
```

### Local Data that requires pre-processing

If your local data requires pre-processing, you can use [`datasets.Dataset.from_dict`](https://huggingface.co/docs/datasets/main/en/package_reference/main_classes#datasets.Dataset.from_dict) to initialize your dataset with a dictionary of lists:

```python
from datasets import Dataset

sentence1_list = []
sentence2_list = []
# Open a file, perform preprocessing, filtering, cleaning, etc.
# and append to the lists

dataset = Dataset.from_dict({
    "sentence1": sentence1_list,

>     "sentence2": sentence2_list,

Reviewer: Maybe use labels more appropriate for column names? It looks (by just looking at the name) that each entry would be a row.

Author: Good call, renamed them in a168c7cee1b40f3c76262badc2af2c019e1d7511

})
```

Each key in the dictionary becomes a column in the resulting dataset.

### Dataset Format

> It's crucial to ensure that your dataset format matches your chosen loss function. This involves checking two things:

Reviewer: Maybe link to the next section.

Author: Good call, resolved in 947ff86ac914967e7b6afed7def28a7a91193173

1. If your loss function requires a *Label* (as indicated in the [Loss Overview](https://sbert.net/docs/sentence_transformer/loss_overview.html) table), your dataset must have a column named **"label"** or **"score"**.
2. All columns other than **"label"** or **"score"** are considered *Inputs* (as indicated in the [Loss Overview](https://sbert.net/docs/sentence_transformer/loss_overview.html) table). The number of these columns must match the number of valid inputs for your chosen loss function. The names of the columns don't matter, **only their order matters**.

> For example, consider a dataset with columns `["text1", "text2", "label"]`, where the `"label"` column contains float similarity scores. This dataset can be used with `CoSENTLoss`, `AnglELoss`, and `CosineSimilarityLoss` because:

Reviewer: ```suggestion
For example, consider a dataset with columns `["text1", "text2", "label"]`, where the `"label"` column contains floating point similarity scores. This dataset can be used with `CoSENTLoss`, `AnglELoss`, and `CosineSimilarityLoss` because:
```

Author: Resolved in 84f301c1915f51f07166bab7f9f228df6978ef6d

## Loss Function

Loss functions measure how well a model performs on a given batch of data and guide the optimization process. The choice of loss function depends on your available data and target task. Refer to the [Loss Overview](https://sbert.net/docs/sentence_transformer/loss_overview.html) for a comprehensive list of options.

Most loss functions can be initialized with just the `SentenceTransformer` you're training and some optional parameters:

```python
from datasets import load_dataset
from sentence_transformers import SentenceTransformer
from sentence_transformers.losses import CoSENTLoss

# Load a model to train/finetune
model = SentenceTransformer("xlm-roberta-base")

# Initialize the CoSENTLoss
# This loss requires pairs of text and a float similarity score as a label
loss = CoSENTLoss(model)

# Load an example training dataset that works with our loss function:

> train_dataset = load_dataset("sentence-transformers/all-nli", "pair-score", split="train")

Reviewer: This dataset does not contain those columns, did you process and upload somewhere else?

Author: The "pair-score" subset does contain those columns. Different loss functions want the data in different formats, so I've released this dataset in a few different subsets.
![image](https://github.com/huggingface/blog/assets/37621491/ecd008d7-3918-4b0d-871c-5ec1cdb1e801)

Reviewer: ah, very cool

Here's an example of how to initialize [`SentenceTransformersTrainingArguments`](https://sbert.net/docs/package_reference/sentence_transformer/training_args.html#sentencetransformertrainingarguments):

```python
from sentence_transformers.training_args import SentenceTransformerTrainingArguments

args = SentenceTransformerTrainingArguments(
    # Required parameter:
    output_dir="models/mpnet-base-all-nli-triplet",
    # Optional training parameters:
    num_train_epochs=1,
    per_device_train_batch_size=16,
    per_device_eval_batch_size=16,
    warmup_ratio=0.1,
    fp16=True,  # Set to False if your GPU can't handle FP16
    bf16=False,  # Set to True if your GPU supports BF16
    batch_sampler=BatchSamplers.NO_DUPLICATES,  # Losses using "in-batch negatives" benefit from no duplicates
    # Optional tracking/debugging parameters:
    eval_strategy="steps",
    eval_steps=100,
    save_strategy="steps",
    save_steps=100,
    save_total_limit=2,
    logging_steps=100,
    run_name="mpnet-base-all-nli-triplet",  # Used in W&B if `wandb` is installed
```

> Note that `eval_strategy` was introduced in `transformers` version `4.41.0`. Prior versions should use `evaluation_strategy` instead.

Reviewer: Is the transformers version not a requirement of the new SentenceTransformers release?

Author: It is, but Sentence Transformers requires `transformers>=4.34.0` currently. I'm hesitant to move to `4.41.0` already, as it's very new (~1 month) and a lot of third party libraries require very specific `transformers` versions. If I increase the minimum version because of a soft deprecation in `transformers`, then I might cause some incompatibilities for users using ST as well as other libraries in the same environment.

Reviewer: Makes sense!

| [`MSEEvaluator`](https://sbert.net/docs/package_reference/sentence_transformer/evaluation.html#mseevaluator) | Source sentences to embed with a teacher model and target sentences to embed with the student model. Can be the same texts. |
| [`ParaphraseMiningEvaluator`](https://sbert.net/docs/package_reference/sentence_transformer/evaluation.html#paraphraseminingevaluator) | Mapping of IDs to sentences & pairs with IDs of duplicate sentences. |
| [`RerankingEvaluator`](https://sbert.net/docs/package_reference/sentence_transformer/evaluation.html#rerankingevaluator) | List of {'query': '..', 'positive': [...], 'negative': [...]} dictionaries. |
| [`TranslationEvaluator`](https://sbert.net/docs/package_reference/sentence_transformer/evaluation.html#translationevaluator) | Pairs of sentences in two separate languages. |
| [`TripletEvaluator`](https://sbert.net/docs/package_reference/sentence_transformer/evaluation.html#tripletevaluator) | (anchor, positive, negative) pairs. |

> Additionally, you can use [`SequentialEvaluator`](https://sbert.net/docs/package_reference/sentence_transformer/evaluation.html#sequentialevaluator) to combine multiple evaluators into one, which can then be passed to the [`SentenceTransformerTrainer`](https://sbert.net/docs/package_reference/sentence_transformer/SentenceTransformer.html#sentence_transformers.SentenceTransformer). The `eval_strategy` and `eval_steps` training arguments determine when the evaluator is run.

Reviewer: I would move the last sentence about `eval_strategy` to the Evaluator intro, as it's not related to `SequentialEvaluator`.

Author: Resolved in a04e407dcc0ec1749038f2178281f9e48f65ef63

If you don't have the necessary evaluation data but still want to track the model's performance on common benchmarks, you can use these evaluators with data from Hugging Face:

### EmbeddingSimilarityEvaluator with STSb

```python
from datasets import load_dataset
from sentence_transformers.evaluation import EmbeddingSimilarityEvaluator, SimilarityFunction

# Load the STSB dataset
eval_dataset = load_dataset("sentence-transformers/stsb", split="validation")

# Initialize the evaluator
dev_evaluator = EmbeddingSimilarityEvaluator(
    sentences1=eval_dataset["sentence1"],
    sentences2=eval_dataset["sentence2"],
    scores=eval_dataset["score"],
    main_similarity=SimilarityFunction.COSINE,
    name="sts-dev",
)
# Run evaluation:
# dev_evaluator(model)

> ```

Reviewer: I'd maybe mention early that the evaluator is passed as another argument to the Trainer for more clarity, even if there's a snippet later.

Reviewer: Also not sure if the `dev_evaluator(model)` gives any info, maybe we could skip it, as evaluation is done by the trainer?

Author: Resolved in 3d68f1fae8a55370004e9b0ed52ce67cf615c05d

eval_dataset = dataset["dev"]
test_dataset = dataset["test"]

# 4. Define a loss function
loss = MultipleNegativesRankingLoss(model)

# 5. (Optional) Specify training arguments
args = SentenceTransformerTrainingArguments(
    # Required parameter:
    output_dir="models/mpnet-base-all-nli-triplet",
    # Optional training parameters:
    num_train_epochs=1,
    per_device_train_batch_size=16,
    per_device_eval_batch_size=16,
    warmup_ratio=0.1,
    fp16=True,  # Set to False if GPU can't handle FP16
    bf16=False,  # Set to True if GPU supports BF16
    batch_sampler=BatchSamplers.NO_DUPLICATES,  # MultipleNegativesRankingLoss benefits from no duplicates
    # Optional tracking/debugging parameters:
    eval_strategy="steps",
    eval_steps=100,
    save_strategy="steps",
    save_steps=100,
    save_total_limit=2,
    logging_steps=100,
    run_name="mpnet-base-all-nli-triplet",  # Used in W&B if `wandb` is installed
)

# 6. (Optional) Create an evaluator & evaluate the base model
dev_evaluator = TripletEvaluator(
    anchors=eval_dataset["anchor"],
    positives=eval_dataset["positive"],
    negatives=eval_dataset["negative"],
    name="all-nli-dev",
)

> dev_evaluator(model)

Reviewer: Is this necessary? Does it record a baseline reference in the training run?

Author: It's not necessary, and yes it does record a baseline reference in the model card. It's quite useful for that, and I recommend it because of it:
![image](https://github.com/huggingface/blog/assets/37621491/24e56cce-0649-4e91-9838-7f48ffb1ddc7)

# 7. Create a trainer & train
trainer = SentenceTransformerTrainer(
    model=model,
    args=args,
    train_dataset=train_dataset,
    eval_dataset=eval_dataset,
    loss=loss,
    evaluator=dev_evaluator,
)
trainer.train()

> # (Optional) Evaluate the trained model on the test set

Reviewer: ```suggestion
# (Optional) Evaluate the trained model on the test set, after training completes
```

    negatives=test_dataset["negative"],
    name="all-nli-test",
)
test_evaluator(model)

# 8. Save the trained model
model.save_pretrained("models/mpnet-base-all-nli-triplet/final")

# 9. (Optional) Push it to the Hugging Face Hub
model.push_to_hub("mpnet-base-all-nli-triplet")
```

In this example I'm finetuning from [`microsoft/mpnet-base`](https://huggingface.co/microsoft/mpnet-base), i.e. a foundational model that is not yet a Sentence Transformer model. This requires more training data than finetuning an existing Sentence Transformer model, like [`all-mpnet-base-v2`](https://huggingface.co/sentence-transformers/all-mpnet-base-v2).

After running this script, the [tomaarsen/mpnet-base-all-nli-triplet](https://huggingface.co/tomaarsen/mpnet-base-all-nli-triplet) model was uploaded for me. The triplet accuracy using cosine similarity, i.e. what percentage of the time `cosine_similarity(anchor, positive) > cosine_similarity(anchor, negative)` is 90.04% for the development set and 91.5% for the testing set! For reference, the [`microsoft/mpnet-base`](https://huggingface.co/microsoft/mpnet-base) model scored only 68.32% on the dev set before training.

> All of this information is stored in the automatically generated model card, including the base model, maximum sequence length, output dimensionality, similarity function, training dataset, language, license, usage, evaluation results, training & evaluation dataset (dataset name, size, approximate statistics, examples), hyperparameters, training logs, environmental impact, training hardware, framework versions, and citations. Without any effort, your uploaded models should contain all the information that your potential users would need to determine whether your model is suitable for them.

Reviewer: Amazing! 🔥

### Callbacks

The Sentence Transformers trainer supports various [`transformers.TrainerCallback`](https://huggingface.co/docs/transformers/main_classes/callback#transformers.TrainerCallback) subclasses, including:

- [`WandbCallback`](https://huggingface.co/docs/transformers/en/main_classes/callback#transformers.integrations.WandbCallback) for logging training metrics to W&B if `wandb` is installed

> - [`TensorBoardCallback`](https://huggingface.co/docs/transformers/en/main_classes/callback#transformers.integrations.TensorBoardCallback) for logging training metrics to TensorBoard if `tensorboard` is accessible

Reviewer: But you don't have to do anything to track using either method, right?

Author: Correct. Clarified in a726511c386752dc2112f06235e90b1b5c2ba811

## Training Examples

The following pages contain training examples with explanations as well as links to code. We recommend that you browse through these to familiarize yourself with the training loop:

* [Semantic Textual Similarity](https://sbert.net/examples/training/sts/README.html)
* [Natural Language Inference](https://sbert.net/examples/training/nli/README.html)
* [Paraphrases](https://sbert.net/examples/training/paraphrases/README.html)
* [Quora Duplicate Questions](https://sbert.net/examples/training/quora_duplicate_questions/README.html)
* [Matryoshka Embeddings](https://sbert.net/examples/training/matryoshka/README.html)
* [Adaptive Layer Models](https://sbert.net/examples/training/adaptive_layer/README.html)
* [Multilingual Models](https://sbert.net/examples/training/multilingual/README.html)
* [Model Distillation](https://sbert.net/examples/training/distillation/README.html)
* [Augmented Sentence Transformers](https://sbert.net/examples/training/data_augmentation/README.html)

> ## Documentation

Reviewer: Maybe add training examples, documentation and the advanced pages inside a common "Additional Resources" section.

Author: I like that, done in 3d7e0d8938efec65eb49d097754f1063c8325165
